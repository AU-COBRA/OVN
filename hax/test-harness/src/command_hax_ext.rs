use lazy_static::{__Deref, lazy_static};
use std::{ffi::OsStr, process::Command};

pub trait CommandHaxExt {
    fn hax<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Self;
}

/// Computes a list of arguments that setup the number of parallel
/// jobs for dune accordingly to environment variable `DUNEJOBS`.
fn dune_jobs_args() -> Vec<String> {
    if let Ok(jobs) = std::env::var("DUNEJOBS") {
        vec!["-j".into(), jobs]
    } else {
        vec![]
    }
}

impl CommandHaxExt for Command {
    fn hax<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Command {
        use assert_cmd::cargo::cargo_bin;
        use std::path::PathBuf;
        struct Paths {
            driver: PathBuf,
            engine: PathBuf,
            cargo_hax: PathBuf,
        }
        const CARGO_HAX: &str = "cargo-hax";
        lazy_static! {
            static ref PATHS: Option<Paths> = {
                if let "yes" | "y" | "true" | "1" = std::env::var("CARGO_TESTS_ASSUME_BUILT").unwrap_or("".into()).to_lowercase().as_str() {
                    return None;
                }
                let root = std::env::current_dir().unwrap();
                let root = root.parent().unwrap();
                let engine_dir = root.join("engine");
                // Make sure binaries are built. Note this doesn't
                // include `hax-engine-names-extract`: its build
                // script requires the driver and CLI of `hax` to be
                // available.
                assert!(Command::new("cargo")
                            .args(&["build", "--bins"])
                            .current_dir(&root)
                            .status()
                            .unwrap()
                            .success());
                let cargo_hax = cargo_bin(CARGO_HAX);
                let driver = cargo_bin("driver-hax-frontend-exporter");
                // Now the driver & CLI are installed, call `cargo
                // build` injecting their paths
                assert!(Command::new("cargo")
                        .args(&["build", "--workspace", "--bin", "hax-engine-names-extract"])
                        .env("HAX_CARGO_COMMAND_PATH", &cargo_hax)
                        .env("HAX_RUSTC_DRIVER_BINARY", &driver)
                        .current_dir(&root)
                        .status()
                        .unwrap()
                        .success());
                assert!(Command::new("dune")
                        .args(&["build"])
                        .args(dune_jobs_args())
                        .env("HAX_JSON_SCHEMA_EXPORTER_BINARY", cargo_bin("hax-export-json-schemas"))
                        .env("HAX_ENGINE_NAMES_EXTRACT_BINARY", cargo_bin("hax-engine-names-extract"))
                        .current_dir(engine_dir.clone())
                        .status()
                        .unwrap()
                        .success());
                Some(Paths {
                    driver,
                    cargo_hax,
                    engine: engine_dir.join("_build/install/default/bin/hax-engine"),
                })
            };
        }
        let mut cmd = match PATHS.deref() {
            Some(paths) => {
                let mut cmd = Command::new(paths.cargo_hax.clone());
                cmd.env("HAX_RUSTC_DRIVER_BINARY", paths.driver.clone());
                cmd.env("HAX_ENGINE_BINARY", paths.engine.clone());
                cmd
            }
            None => Command::new(CARGO_HAX),
        };
        cmd.args(args);
        // As documented in
        // https://doc.rust-lang.org/cargo/reference/environment-variables.html#dynamic-library-paths,
        // [cargo run] (and thus also [cargo test]) sets dynamic
        // library paths, which causes some issues with dependencies
        // when compiling without rustup
        for env in ["DYLD_FALLBACK_LIBRARY_PATH", "LD_LIBRARY_PATH"] {
            cmd.env_remove(env);
        }
        cmd
    }
}
