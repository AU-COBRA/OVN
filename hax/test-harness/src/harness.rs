mod command_hax_ext;
use command_hax_ext::*;
use serde_json::{Map, Value};
use std::process::{Command, Stdio};

#[derive(Clone, Debug, serde::Serialize)]
pub enum TestKind {
    Translate { backend: String },
}

impl TestKind {
    fn as_name(&self) -> String {
        (match self {
            TestKind::Translate { backend } => ["into".to_string(), backend.clone()],
        })
        .join("-")
    }
}

#[allow(dead_code)]
fn bool_true() -> bool {
    true
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct TestSnapshot {
    #[serde(default = "bool_true")]
    pub stderr: bool,
    #[serde(default = "bool_true")]
    pub stdout: bool,
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct TestSpec {
    /// is the test optional? (useful for slow tests for instance)
    pub optional: bool,
    /// a broken test a test that should succeed (or fail) but does
    /// not dues to a bug to be fixed (see field [issue_id] below)
    pub broken: bool,
    /// Github issue ID
    pub issue_id: Option<u64>,
    /// Is that a positive or a negative test?
    pub positive: bool,
    pub snapshot: TestSnapshot,
    pub include_flag: Option<String>,
    pub backend_options: Option<Vec<String>>,
}

impl From<Value> for TestSpec {
    /// Parse a JSON value into a TestSpec
    fn from(o: Value) -> Self {
        fn as_opt_bool(v: &Value, def: bool) -> Option<bool> {
            if v.is_null() {
                return Some(def);
            }
            v.as_bool()
        }
        fn as_bool(o: &Value, k: &str, def: bool) -> bool {
            let v = &o[k];
            as_opt_bool(v, def)
                .expect(format!("[{}] was expected to be a boolean, got {}", k, v).as_str())
        }
        let snapshot = &o["snapshot"];
        TestSpec {
            optional: as_bool(&o, "optional", false),
            broken: as_bool(&o, "broken", false),
            positive: as_bool(&o, "positive", true),
            issue_id: o["positive"].as_u64(),
            include_flag: o["include-flag"].as_str().map(|s| s.into()),
            backend_options: serde_json::from_value(o["backend-options"].clone()).unwrap(),
            snapshot: as_opt_bool(snapshot, true)
                .map(|b| TestSnapshot {
                    stderr: b,
                    stdout: b,
                })
                .or_else(|| match snapshot.as_str() {
                    Some(v @ ("stdout" | "stderr" | "both" | "none")) => Some(TestSnapshot {
                        stdout: matches!(v, "stdout" | "both"),
                        stderr: matches!(v, "stderr" | "both"),
                    }),
                    Some(v) => panic!(
                        "[snapshot] is \"{}\" but was expected to be \"stderr\", \"stdout\" or \"both\"", v
                    ),
                    None => None,
                })
                .unwrap_or_else(|| TestSnapshot {
                    stderr: as_bool(&snapshot, "stderr", true),
                    stdout: as_bool(&snapshot, "stdout", true),
                }),
        }
    }
}

/// The information for a test is given by `cargo metadata`
#[derive(Clone, Debug, serde::Serialize)]
pub struct TestInfo {
    pub name: String,
    pub manifest: std::path::PathBuf,
    pub description: Option<String>,
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct Test {
    pub kind: TestKind,
    pub info: TestInfo,
    pub spec: TestSpec,
}

impl Test {
    fn as_args(&self) -> Vec<String> {
        match &self.kind {
            TestKind::Translate { backend } => {
                let mut args = vec![];
                args.push("into".to_string());
                if let Some(i) = self.spec.include_flag.as_ref() {
                    args.push("-i".to_string());
                    args.push(i.to_string());
                }
                args.push("--dry-run".to_string());
                args.push(backend.clone());
                if let Some(backend_options) = &self.spec.backend_options {
                    args.extend_from_slice(backend_options.clone().as_slice());
                }
                args
            }
        }
    }
}

impl std::fmt::Display for Test {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{} - {:?}", self.info.name, self.kind)?;
        if let Some(issue_id) = self.spec.issue_id {
            write!(f, " #{}", issue_id)?;
        };
        Ok(())
    }
}

impl Test {
    fn into_runner(self, workspace: String) -> Result<(), libtest_mimic::Failed> {
        // 1. cook a command
        let mut cmd = Command::hax(&["-C"]);
        cmd.arg("--manifest-path").arg(self.info.manifest.clone());
        cmd.arg(";");
        cmd.stdout(Stdio::piped()).stderr(Stdio::piped());
        cmd.args(self.as_args());

        // 2. execute it (twice, idea of @franziskuskiefer, so that
        // the messages related to building dependencies are not
        // included in the second one)
        let _ = cmd.output().unwrap();
        let out = cmd.output().unwrap();

        let command_successful = out.status.success();
        let cleanup = |s: String| {
            use lazy_static::lazy_static;
            use regex::Regex;
            lazy_static! {
                // Regex [TIME] matches compile times
                static ref TIME: Regex = Regex::new(r"\bin \d+(\.\d+)?s\b").unwrap();
                static ref LOCK: Regex = Regex::new(r"Blocking waiting for \w+ lock on (the registry index|build directory|package cache)").unwrap();
            }
            TIME.replace_all(
                LOCK.replace_all(
                    &s.replace(r"\", "/").replace(&workspace, "WORKSPACE_ROOT"),
                    "",
                )
                .as_ref(),
                "in XXs",
            )
            .trim()
            .to_string()
        };
        let serr = cleanup(String::from_utf8_lossy(&out.stderr).to_string());
        let sout = String::from_utf8_lossy(&out.stdout).to_string();

        // 3. make sure the test is successful
        let mut snapshot: Map<String, Value> = Map::new();
        if self.spec.snapshot.stderr {
            snapshot.insert("stderr".to_string(), Value::String(serr.clone()));
        }
        if self.spec.snapshot.stdout {
            snapshot.insert(
                "stdout".to_string(),
                serde_json::from_str(&sout)
                    .map(|out: hax_types::engine_api::Output| {
                        use serde_json::json;
                        json!({
                            "diagnostics": Value::Array(out.diagnostics.into_iter().map(|diag| json!({
                                "spans": Value::Array(diag.span.clone().into_iter().map(|span| Value::String(format!("{:?}", span))).collect()),
                                "message": Value::String(format!("{}", diag)),
                            })).collect()),
                            "files": Value::Object(out.files.into_iter().map(|file| (file.path, Value::String(file.contents))).collect())
                        })
                    })
                    .unwrap_or_else(|_| Value::String(cleanup(sout.clone()))),
            );
        }

        if !snapshot.is_empty() {
            let exit = out.status.code().unwrap_or(std::i32::MAX);
            snapshot.insert("exit".to_string(), exit.into());
            let snapshot = Value::Object(snapshot);
            let name = format!("{} {}", self.info.name, self.kind.as_name());

            let mut info = self.clone();
            info.info.manifest = info.info.manifest.strip_prefix(workspace).unwrap().into();
            insta::with_settings!({
                info => &info,
            }, { insta::assert_toml_snapshot!(name, snapshot) })
        }

        let err = |s: &str| {
            Err(format!(
                "Command {s}.\nThe command was: {:?}{}",
                cmd,
                if command_successful {
                    "".to_string()
                } else {
                    format!("\nSTDOUT:\n{}\nSTDERR:\n{}", sout, serr)
                }
            ))
        };
        match (command_successful, (self.spec.positive, self.spec.broken)) {
            (true, (true, false) | (false, true)) => Ok(()),
            (false, (false, false) | (true, true)) => Ok(()),
            (false, (false, true)) => err("failed, but this is a negative test marked broken")?,
            (false, (true, false)) => err("failed")?,
            (true, (true, true)) => err("succeeded, but this is a positive test marked broken")?,
            (true, (false, false)) => err("succeeded, but this is a negative test")?,
        }
    }

    fn into_trial(&self, workspace: &String) -> libtest_mimic::Trial {
        libtest_mimic::Trial::test(format!("{}", &self), {
            let test = self.clone();
            let workspace = workspace.clone();
            move || test.clone().into_runner(workspace)
        })
        .with_kind(if self.spec.positive {
            "positive"
        } else {
            "negative"
        })
        .with_ignored_flag(self.spec.optional)
    }
}

/// Given [metadata] the table declared in a test's [Cargo.toml]
/// [workspace.hax-tests], this function returns a list of tests
fn parse_hax_tests_metadata(info: TestInfo, metadata: &Value) -> Vec<Test> {
    if metadata.is_null() {
        return vec![];
    }

    metadata
        .as_object()
        .expect(
            format!(
                "Expected value at key [hax-tests] to be a dictionary for package {:#?}",
                info
            )
            .as_str(),
        )
        .into_iter()
        .flat_map(|(a, o)| {
            o.as_object()
                .expect(
                    format!(
                        "Expected value at key [{}] be a dictionary for package {:#?}",
                        a, info
                    )
                    .as_str(),
                )
                .into_iter()
                .flat_map(|(key, o)| key.split("+").map(|k| (k.trim().to_string(), o.clone())))
                .map(|(b, o)| (a.clone(), b, o))
        })
        .map(|(a, b, o)| Test {
            spec: o.into(),
            info: info.clone(),
            kind: match a.as_str() {
                "into" => TestKind::Translate { backend: b },
                _ => panic!(
                    "unexpected metadata [hax-tests.{}.{}] for package {:#?}",
                    a, b, info
                ),
            },
        })
        .collect()
}

fn main() {
    let metadata = cargo_metadata::MetadataCommand::new()
        .manifest_path("../tests/Cargo.toml")
        .exec()
        .unwrap();
    let workspace_root: String = metadata.workspace_root.into();

    let mut args = libtest_mimic::Arguments::from_args();
    args.test_threads = Some(1);

    libtest_mimic::run(
        &args,
        metadata
            .packages
            .into_iter()
            .flat_map(|o| {
                parse_hax_tests_metadata(
                    TestInfo {
                        name: o.name,
                        description: o.description,
                        manifest: o.manifest_path.into(),
                    },
                    &o.metadata["hax-tests"],
                )
            })
            .map(|test| test.into_trial(&workspace_root))
            .collect(),
    )
    .exit();
}
