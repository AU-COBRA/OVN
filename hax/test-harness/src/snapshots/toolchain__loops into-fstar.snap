---
source: test-harness/src/harness.rs
expression: snapshot
info:
  kind:
    Translate:
      backend: fstar
  info:
    name: loops
    manifest: loops/Cargo.toml
    description: ~
  spec:
    optional: false
    broken: false
    issue_id: ~
    positive: true
    snapshot:
      stderr: true
      stdout: true
    include_flag: ~
    backend_options: ~
---
exit = 0
stderr = 'Finished `dev` profile [unoptimized + debuginfo] target(s) in XXs'

[stdout]
diagnostics = []

[stdout.files]
"Loops.For_loops.fst" = '''
module Loops.For_loops
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

let bool_returning (x: u8) : bool = x <. 10uy

let chunks (v_CHUNK_LEN: usize) (arr: Alloc.Vec.t_Vec usize Alloc.Alloc.t_Global) : usize =
  let acc:usize = sz 0 in
  let chunks:Core.Slice.Iter.t_ChunksExact usize =
    Core.Slice.impl__chunks_exact #usize
      (Core.Ops.Deref.f_deref #(Alloc.Vec.t_Vec usize Alloc.Alloc.t_Global)
          #FStar.Tactics.Typeclasses.solve
          arr
        <:
        t_Slice usize)
      v_CHUNK_LEN
  in
  let acc:usize =
    Core.Iter.Traits.Iterator.f_fold (Core.Iter.Traits.Collect.f_into_iter #(Core.Slice.Iter.t_ChunksExact
            usize)
          #FStar.Tactics.Typeclasses.solve
          (Core.Clone.f_clone #(Core.Slice.Iter.t_ChunksExact usize)
              #FStar.Tactics.Typeclasses.solve
              chunks
            <:
            Core.Slice.Iter.t_ChunksExact usize)
        <:
        Core.Slice.Iter.t_ChunksExact usize)
      acc
      (fun acc chunk ->
          let acc:usize = acc in
          let chunk:t_Slice usize = chunk in
          let mean:usize = sz 0 in
          let mean:usize =
            Core.Iter.Traits.Iterator.f_fold (Core.Iter.Traits.Collect.f_into_iter #(t_Slice usize)
                  #FStar.Tactics.Typeclasses.solve
                  chunk
                <:
                Core.Slice.Iter.t_Iter usize)
              mean
              (fun mean item ->
                  let mean:usize = mean in
                  let item:usize = item in
                  mean +! item <: usize)
          in
          let acc:usize = acc +! (mean /! v_CHUNK_LEN <: usize) in
          acc)
  in
  let acc:usize =
    Core.Iter.Traits.Iterator.f_fold (Core.Iter.Traits.Collect.f_into_iter #(t_Slice usize)
          #FStar.Tactics.Typeclasses.solve
          (Core.Slice.Iter.impl_88__remainder #usize chunks <: t_Slice usize)
        <:
        Core.Slice.Iter.t_Iter usize)
      acc
      (fun acc item ->
          let acc:usize = acc in
          let item:usize = item in
          acc -! item <: usize)
  in
  acc

let composed_range (n: usize) : usize =
  let acc:usize = sz 0 in
  let acc:usize =
    Core.Iter.Traits.Iterator.f_fold (Core.Iter.Traits.Collect.f_into_iter #(Core.Iter.Adapters.Chain.t_Chain
              (Core.Ops.Range.t_Range usize) (Core.Ops.Range.t_Range usize))
          #FStar.Tactics.Typeclasses.solve
          (Core.Iter.Traits.Iterator.f_chain #(Core.Ops.Range.t_Range usize)
              #FStar.Tactics.Typeclasses.solve
              #(Core.Ops.Range.t_Range usize)
              ({ Core.Ops.Range.f_start = sz 0; Core.Ops.Range.f_end = n }
                <:
                Core.Ops.Range.t_Range usize)
              ({
                  Core.Ops.Range.f_start = n +! sz 10 <: usize;
                  Core.Ops.Range.f_end = n +! sz 50 <: usize
                }
                <:
                Core.Ops.Range.t_Range usize)
            <:
            Core.Iter.Adapters.Chain.t_Chain (Core.Ops.Range.t_Range usize)
              (Core.Ops.Range.t_Range usize))
        <:
        Core.Iter.Adapters.Chain.t_Chain (Core.Ops.Range.t_Range usize)
          (Core.Ops.Range.t_Range usize))
      acc
      (fun acc i ->
          let acc:usize = acc in
          let i:usize = i in
          (acc +! i <: usize) +! sz 1 <: usize)
  in
  acc

let enumerate_chunks (arr: Alloc.Vec.t_Vec usize Alloc.Alloc.t_Global) : usize =
  let acc:usize = sz 0 in
  let acc:usize =
    Core.Iter.Traits.Iterator.f_fold (Core.Iter.Traits.Collect.f_into_iter #(Core.Iter.Adapters.Enumerate.t_Enumerate
            (Core.Slice.Iter.t_Chunks usize))
          #FStar.Tactics.Typeclasses.solve
          (Core.Iter.Traits.Iterator.f_enumerate #(Core.Slice.Iter.t_Chunks usize)
              #FStar.Tactics.Typeclasses.solve
              (Core.Slice.impl__chunks #usize
                  (Core.Ops.Deref.f_deref #(Alloc.Vec.t_Vec usize Alloc.Alloc.t_Global)
                      #FStar.Tactics.Typeclasses.solve
                      arr
                    <:
                    t_Slice usize)
                  (sz 4)
                <:
                Core.Slice.Iter.t_Chunks usize)
            <:
            Core.Iter.Adapters.Enumerate.t_Enumerate (Core.Slice.Iter.t_Chunks usize))
        <:
        Core.Iter.Adapters.Enumerate.t_Enumerate (Core.Slice.Iter.t_Chunks usize))
      acc
      (fun acc temp_1_ ->
          let acc:usize = acc in
          let i, chunk:(usize & t_Slice usize) = temp_1_ in
          Rust_primitives.Hax.Folds.fold_enumerated_slice chunk
            (fun acc temp_1_ ->
                let acc:usize = acc in
                let _:usize = temp_1_ in
                true)
            acc
            (fun acc temp_1_ ->
                let acc:usize = acc in
                let j, x:(usize & usize) = temp_1_ in
                (i +! j <: usize) +! x <: usize)
          <:
          usize)
  in
  acc

let f (_: Prims.unit) : u8 =
  let acc:u8 = 0uy in
  Rust_primitives.Hax.Folds.fold_range 1uy
    10uy
    (fun acc temp_1_ ->
        let acc:u8 = acc in
        let _:u8 = temp_1_ in
        true)
    acc
    (fun acc i ->
        let acc:u8 = acc in
        let i:u8 = i in
        let acc:u8 = acc +! i in
        let _:bool = bool_returning i in
        acc)

let iterator (arr: Alloc.Vec.t_Vec usize Alloc.Alloc.t_Global) : usize =
  let acc:usize = sz 0 in
  let acc:usize =
    Core.Iter.Traits.Iterator.f_fold (Core.Iter.Traits.Collect.f_into_iter #(Core.Slice.Iter.t_Iter
            usize)
          #FStar.Tactics.Typeclasses.solve
          (Core.Slice.impl__iter #usize
              (Core.Ops.Deref.f_deref #(Alloc.Vec.t_Vec usize Alloc.Alloc.t_Global)
                  #FStar.Tactics.Typeclasses.solve
                  arr
                <:
                t_Slice usize)
            <:
            Core.Slice.Iter.t_Iter usize)
        <:
        Core.Slice.Iter.t_Iter usize)
      acc
      (fun acc item ->
          let acc:usize = acc in
          let item:usize = item in
          acc +! item <: usize)
  in
  acc

let nested (arr: Alloc.Vec.t_Vec usize Alloc.Alloc.t_Global) : usize =
  let acc:usize = sz 0 in
  let acc:usize =
    Core.Iter.Traits.Iterator.f_fold (Core.Iter.Traits.Collect.f_into_iter #(Core.Slice.Iter.t_Iter
            usize)
          #FStar.Tactics.Typeclasses.solve
          (Core.Slice.impl__iter #usize
              (Core.Ops.Deref.f_deref #(Alloc.Vec.t_Vec usize Alloc.Alloc.t_Global)
                  #FStar.Tactics.Typeclasses.solve
                  arr
                <:
                t_Slice usize)
            <:
            Core.Slice.Iter.t_Iter usize)
        <:
        Core.Slice.Iter.t_Iter usize)
      acc
      (fun acc item ->
          let acc:usize = acc in
          let item:usize = item in
          Core.Iter.Traits.Iterator.f_fold (Core.Iter.Traits.Collect.f_into_iter #(Core.Iter.Adapters.Rev.t_Rev
                  (Core.Ops.Range.t_Range usize))
                #FStar.Tactics.Typeclasses.solve
                (Core.Iter.Traits.Iterator.f_rev #(Core.Ops.Range.t_Range usize)
                    #FStar.Tactics.Typeclasses.solve
                    ({ Core.Ops.Range.f_start = sz 0; Core.Ops.Range.f_end = item }
                      <:
                      Core.Ops.Range.t_Range usize)
                  <:
                  Core.Iter.Adapters.Rev.t_Rev (Core.Ops.Range.t_Range usize))
              <:
              Core.Iter.Adapters.Rev.t_Rev (Core.Ops.Range.t_Range usize))
            acc
            (fun acc i ->
                let acc:usize = acc in
                let i:usize = i in
                let acc:usize = acc +! sz 1 in
                Core.Iter.Traits.Iterator.f_fold (Core.Iter.Traits.Collect.f_into_iter #(Core.Iter.Adapters.Zip.t_Zip
                          (Core.Slice.Iter.t_Iter usize) (Core.Ops.Range.t_Range usize))
                      #FStar.Tactics.Typeclasses.solve
                      (Core.Iter.Traits.Iterator.f_zip #(Core.Slice.Iter.t_Iter usize)
                          #FStar.Tactics.Typeclasses.solve
                          #(Core.Ops.Range.t_Range usize)
                          (Core.Slice.impl__iter #usize
                              (Core.Ops.Deref.f_deref #(Alloc.Vec.t_Vec usize Alloc.Alloc.t_Global)
                                  #FStar.Tactics.Typeclasses.solve
                                  arr
                                <:
                                t_Slice usize)
                            <:
                            Core.Slice.Iter.t_Iter usize)
                          ({ Core.Ops.Range.f_start = sz 4; Core.Ops.Range.f_end = i }
                            <:
                            Core.Ops.Range.t_Range usize)
                        <:
                        Core.Iter.Adapters.Zip.t_Zip (Core.Slice.Iter.t_Iter usize)
                          (Core.Ops.Range.t_Range usize))
                    <:
                    Core.Iter.Adapters.Zip.t_Zip (Core.Slice.Iter.t_Iter usize)
                      (Core.Ops.Range.t_Range usize))
                  acc
                  (fun acc j ->
                      let acc:usize = acc in
                      let j:(usize & usize) = j in
                      (((acc +! item <: usize) +! i <: usize) +! j._1 <: usize) +! j._2 <: usize))
          <:
          usize)
  in
  acc

let pattern (arr: Alloc.Vec.t_Vec (usize & usize) Alloc.Alloc.t_Global) : usize =
  let acc:usize = sz 0 in
  let acc:usize =
    Core.Iter.Traits.Iterator.f_fold (Core.Iter.Traits.Collect.f_into_iter #(Alloc.Vec.t_Vec
              (usize & usize) Alloc.Alloc.t_Global)
          #FStar.Tactics.Typeclasses.solve
          arr
        <:
        Alloc.Vec.Into_iter.t_IntoIter (usize & usize) Alloc.Alloc.t_Global)
      acc
      (fun acc temp_1_ ->
          let acc:usize = acc in
          let x, y:(usize & usize) = temp_1_ in
          acc +! (x *! y <: usize) <: usize)
  in
  acc

let range1 (_: Prims.unit) : usize =
  let acc:usize = sz 0 in
  let acc:usize =
    Rust_primitives.Hax.Folds.fold_range (sz 0)
      (sz 15)
      (fun acc temp_1_ ->
          let acc:usize = acc in
          let _:usize = temp_1_ in
          true)
      acc
      (fun acc i ->
          let acc:usize = acc in
          let i:usize = i in
          acc +! i <: usize)
  in
  acc

let range2 (n: usize) : usize =
  let acc:usize = sz 0 in
  let acc:usize =
    Rust_primitives.Hax.Folds.fold_range (sz 0)
      (n +! sz 10 <: usize)
      (fun acc temp_1_ ->
          let acc:usize = acc in
          let _:usize = temp_1_ in
          true)
      acc
      (fun acc i ->
          let acc:usize = acc in
          let i:usize = i in
          (acc +! i <: usize) +! sz 1 <: usize)
  in
  acc

let rev_range (n: usize) : usize =
  let acc:usize = sz 0 in
  let acc:usize =
    Core.Iter.Traits.Iterator.f_fold (Core.Iter.Traits.Collect.f_into_iter #(Core.Iter.Adapters.Rev.t_Rev
            (Core.Ops.Range.t_Range usize))
          #FStar.Tactics.Typeclasses.solve
          (Core.Iter.Traits.Iterator.f_rev #(Core.Ops.Range.t_Range usize)
              #FStar.Tactics.Typeclasses.solve
              ({ Core.Ops.Range.f_start = sz 0; Core.Ops.Range.f_end = n }
                <:
                Core.Ops.Range.t_Range usize)
            <:
            Core.Iter.Adapters.Rev.t_Rev (Core.Ops.Range.t_Range usize))
        <:
        Core.Iter.Adapters.Rev.t_Rev (Core.Ops.Range.t_Range usize))
      acc
      (fun acc i ->
          let acc:usize = acc in
          let i:usize = i in
          (acc +! i <: usize) +! sz 1 <: usize)
  in
  acc
'''
"Loops.Recognized_loops.fst" = '''
module Loops.Recognized_loops
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

let enumerated_chunked_slice (#v_T: Type0) (slice: t_Slice v_T) : u64 =
  let count:u64 = 0uL in
  Rust_primitives.Hax.Folds.fold_enumerated_chunked_slice (sz 3)
    slice
    (fun count i ->
        let count:u64 = count in
        let i:usize = i in
        i <= Core.Slice.impl__len #v_T slice)
    count
    (fun count i ->
        let count:u64 = count in
        let i:(usize & t_Slice v_T) = i in
        let count:u64 = count +! 3uL in
        count)

let enumerated_slice (#v_T: Type0) (slice: t_Slice v_T) : u64 =
  let count:u64 = 0uL in
  Rust_primitives.Hax.Folds.fold_enumerated_slice slice
    (fun count i ->
        let count:u64 = count in
        let i:usize = i in
        i <=. sz 10 <: bool)
    count
    (fun count i ->
        let count:u64 = count in
        let i:(usize & v_T) = i in
        let count:u64 = count +! 2uL in
        count)

let range (_: Prims.unit) : u64 =
  let count:u64 = 0uL in
  Rust_primitives.Hax.Folds.fold_range 0uy
    10uy
    (fun count i ->
        let count:u64 = count in
        let i:u8 = i in
        i <=. 10uy <: bool)
    count
    (fun count i ->
        let count:u64 = count in
        let i:u8 = i in
        let count:u64 = count +! 1uL in
        count)

let range_step_by (_: Prims.unit) : u64 =
  let count:u64 = 0uL in
  Rust_primitives.Hax.Folds.fold_range_step_by 0uy
    10uy
    (sz 2)
    (fun count i ->
        let count:u64 = count in
        let i:u8 = i in
        i <=. 10uy <: bool)
    count
    (fun count i ->
        let count:u64 = count in
        let i:u8 = i in
        let count:u64 = count +! 1uL in
        count)
'''
"Loops.While_loops.fst" = '''
module Loops.While_loops
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

let f (_: Prims.unit) : u8 =
  let x:u8 = 0uy in
  let x:u8 =
    Rust_primitives.f_while_loop (fun x ->
          let x:u8 = x in
          x <. 10uy <: bool)
      x
      (fun x ->
          let x:u8 = x in
          let x:u8 = x +! 3uy in
          x)
  in
  x +! 12uy
'''
